src/frequenz/dispatch/actor.py

```diff
@@ -0,0 +1,279 @@
+
+    async def _run(self) -> None:
+        """Run the actor."""
+        try:
+            while True:
+                await self._fetch()
+                await asyncio.sleep(self._poll_interval.total_seconds())
+        except asyncio.CancelledError:
+            for task in self._scheduled.values():
+                task.cancel()
+            raise
+
+    async def _fetch(self) -> None:
+        """Fetch all relevant dispatches."""
+        old_dispatches = self._dispatches
+        self._dispatches = {}
+
+        try:
+            _logger.info("Fetching dispatches for microgrid %s", self._microgrid_id)
+            async for dispatch in self._client.list(microgrid_id=self._microgrid_id):
+                self._dispatches[dispatch.id] = dispatch
+
+                old_dispatch = old_dispatches.pop(dispatch.id, None)
+                if not old_dispatch:
+                    self._update_dispatch_schedule(dispatch, None)
+                    _logger.info("New dispatch: %s", dispatch)
+                    await self._updated_dispatch_sender.send(Created(dispatch=dispatch))
+                elif dispatch.update_time != old_dispatch.update_time:
+                    self._update_dispatch_schedule(dispatch, old_dispatch)
+                    _logger.info("Updated dispatch: %s", dispatch)
+                    await self._updated_dispatch_sender.send(Updated(dispatch=dispatch))
+
+        except grpc.aio.AioRpcError as error:
+            _logger.error("Error fetching dispatches: %s", error)
+            self._dispatches = old_dispatches
+            return
+
+        for dispatch in old_dispatches.values():
+            _logger.info("Deleted dispatch: %s", dispatch)
+            await self._updated_dispatch_sender.send(Deleted(dispatch=dispatch))
+            if task := self._scheduled.pop(dispatch.id, None):
+                task.cancel()
+
+    def _update_dispatch_schedule(
+        self, dispatch: Dispatch, old_dispatch: Dispatch | None
+    ) -> None:
+        """Update the schedule for a dispatch.
+
+        Schedules, reschedules or cancels the dispatch based on the start_time
+        and active status.
+
+        For example:
+            * when the start_time changes, the dispatch is rescheduled
+            * when the dispatch is deactivated, the dispatch is cancelled
+
+        Args:
+            dispatch: The dispatch to update the schedule for.
+            old_dispatch: The old dispatch, if available.
+        """
+        if (
+            old_dispatch
+            and old_dispatch.active
+            and old_dispatch.start_time != dispatch.start_time
+        ):
+            if task := self._scheduled.pop(dispatch.id, None):
+                task.cancel()
+
+        if dispatch.active and dispatch.id not in self._scheduled:
+            self._scheduled[dispatch.id] = asyncio.create_task(
+                self._schedule_task(dispatch)
+            )
+
+    async def _schedule_task(self, dispatch: Dispatch) -> None:
```

Comment:
My mental picture of how this should have worked was, instead of having one task per dispatch, to just have one task that calculates the next run for all dispatches and task the one that needs to run next, then sleeps only for that time, sends the ready dispatch through the channel, and calculates the next dispatch to be run again.

I think this scheme should be simpler and more memory efficient, as you don't need one task per dispatch. Do you see any advantages with the current approach?

I guess it might be a bit more CPU efficient, as the next dispatch for all dispatches doesn't need to be calculated every time a dispatch is ready, but this could be optimized fairly easily by saving all the next run calculation for all dispatches in a sorted list and only calculate the next run for the dispatch that was just ready. Then we just need to invalidate this list each time dispatches are changed.

Reply:
This one might be too much work to implement, so I think it is better to discuss it first, and probably it is better to sense the general feeling here first, if it's not a trivial "let's go with this approach" (or let's NOT) before creating an issue, so please go ahead and leave comments here.

Reply:
I think my only concern with your suggestion is that it might accidentally skip dispatches that are very close together, though I can't come up for a good explanation why it would, just an .. gut feeling to pay extra attention on that area when implementing it..

I was also wondering if the awaiting for sending might take a bit of time and thus delays the notification for the next very close dispatch to be executed. What if we waited longer than the next dispatches scheduled time, how do we make sure we don't skip it?

Reply:
> I think my only concern with your suggestion is that it might accidentally skip dispatches that are very close together, though I can't come up for a good explanation why it would, just an .. gut feeling to pay extra attention on that area when implementing it..

If done properly, I don't think it should. I didn't invented this, I've seen this (or a similar approach) suggested in [libev](http://software.schmorp.de/pkg/libev.html) as a [very efficient way to implement timers](http://pod.tst.eu/http://cvs.schmorp.de/libev/ev.pod#code_ev_timer_code_relative_and_opti) (see point 4).

> I was also wondering if the awaiting for sending might take a bit of time and thus delays the notification for the next very close dispatch to be executed. What if we waited longer than the next dispatches scheduled time, how do we make sure we don't skip it?

If we await on the send takes a long time, I guess the next notification will be delayed too as it will use the same channel to notify, right? We could do the send in a separate task but then we end up with many tasks again, but I think sends will be serialized in any case because they operate in the same channel, so using a different task is probably not having any real effect.